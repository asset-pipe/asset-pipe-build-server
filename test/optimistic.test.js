'use strict';

const publishAssets = require('../lib/publish-assets');
const publishInstructions = require('../lib/publish-instructions');
const Sink = require('@asset-pipe/sink-mem');
const express = require('express');
const supertest = require('supertest');
const Router = require('../lib/main');
const { endWorkers } = require('../lib/utils');

function createTestServerFor(router) {
    const app = express();
    app.use(router);
    return new Promise(resolve => {
        const server = app.listen(() => {
            resolve({
                server,
                port: server.address().port,
            });
        });
    });
}

test('integration - optimistic bundling of js feeds', async () => {
    const sink = new Sink();

    await publishAssets(sink, {
        tag: 'podlet1',
        type: 'js',
        data: [
            {
                id: 'c645cf572a8f5acf8716e4846b408d3b1ca45c58',
                source:
                    '"use strict";module.exports.world=function(){return"world"};',
                deps: {},
                file: './assets/js/bar.js',
            },
        ],
    });

    await publishInstructions(sink, {
        tag: 'layout1',
        type: 'js',
        data: ['podlet1'],
    });

    expect(sink.db).toMatchSnapshot();
});

test('integration - publish instructions before publishing assets', async () => {
    const sink = new Sink();

    await publishInstructions(sink, {
        tag: 'layout1',
        type: 'js',
        data: ['podlet1'],
    });

    await publishAssets(sink, {
        tag: 'podlet1',
        type: 'js',
        data: [
            {
                id: 'c645cf572a8f5acf8716e4846b408d3b1ca45c58',
                source:
                    '"use strict";module.exports.world=function(){return"world"};',
                deps: {},
                file: './assets/js/bar.js',
            },
        ],
    });

    expect(sink.db).toMatchSnapshot();
});

test('integration - optimistic bundling of css feeds', async () => {
    const sink = new Sink();

    await publishAssets(sink, {
        tag: 'podlet1',
        type: 'css',
        data: [
            {
                id: 'c645cf572a8f5acf8716e4846b408d3b1ca45c58',
                source: 'h1 { color: red; }',
                deps: {},
                file: './assets/css/bar.js',
            },
        ],
    });

    await publishInstructions(sink, {
        tag: 'layout1',
        type: 'css',
        data: ['podlet1'],
    });

    expect(sink.db).toMatchSnapshot();
});

test('integration - publishing a feed via the /publish-assets endpoint', async () => {
    expect.assertions(1);
    const router = new Router();
    const { server } = await createTestServerFor(router.router());
    const post = supertest(server).post;
    const payload = {
        tag: 'podlet1',
        type: 'js',
        data: [
            {
                id: 'c645cf572a8f5acf8716e4846b408d3b1ca45c58',
                source:
                    '"use strict";module.exports.world=function(){return"world"};',
                deps: {},
                file: './assets/js/bar.js',
            },
        ],
    };
    const { body } = await post('/publish-assets')
        .send(payload)
        .expect(200);
    expect(body).toMatchSnapshot();
    return server.close();
});

test('integration - publishing bundling instructions via the /publish-instructions endpoint', async () => {
    expect.assertions(1);
    const router = new Router();
    const { server } = await createTestServerFor(router.router());
    const post = supertest(server).post;
    await post('/publish-assets').send({
        tag: 'podlet1',
        type: 'js',
        data: [
            {
                id: 'c645cf572a8f5acf8716e4846b408d3b1ca45c58',
                source:
                    '"use strict";module.exports.world=function(){return"world"};',
                deps: {},
                file: './assets/js/bar.js',
            },
        ],
    });
    const { body } = await post('/publish-instructions')
        .send({ tag: 'layout1', type: 'js', data: ['podlet1'] })
        .expect(200);

    expect(body.success).toBe(true);
    return server.close();
});

afterAll(() => endWorkers());
