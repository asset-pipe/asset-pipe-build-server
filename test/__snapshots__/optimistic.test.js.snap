// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration - optimistic bundling of css feeds 1`] = `
Object {
  "2c408796d55d03d6fbaef774c591b868cf07cac4147bc61d3b40825e16f725ab.css": "\\"h1 { color: red; }\\"",
  "meta/instructions.json": "{
  \\"layout1/css\\": {
    \\"tag\\": \\"layout1\\",
    \\"type\\": \\"css\\",
    \\"data\\": [
      \\"podlet1\\"
    ]
  }
}",
  "meta/tags.json": "{
  \\"podlet1/css\\": \\"ed8954cab48b223649511da5e5d51f1cfce3f4056bd11513be00a60f338cfa27\\"
}",
  "podlet1-ed8954cab48b223649511da5e5d51f1cfce3f4056bd11513be00a60f338cfa27.json": "[
  {
    \\"id\\": \\"c645cf572a8f5acf8716e4846b408d3b1ca45c58\\",
    \\"source\\": \\"h1 { color: red; }\\",
    \\"deps\\": {},
    \\"file\\": \\"./assets/css/bar.js\\"
  }
]",
}
`;

exports[`integration - optimistic bundling of js feeds 1`] = `
Object {
  "8b9a3142fa84751f4a0b6578f58024bd3b98aea6fb47f8e87dd0c813a250fa95.js": "\\"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\\\\\"function\\\\\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\\\\\"Cannot find module '\\\\\\"+o+\\\\\\"'\\\\\\");throw f.code=\\\\\\"MODULE_NOT_FOUND\\\\\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\\\\\"function\\\\\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({},{},[]);\\\\n\\"",
  "meta/instructions.json": "{
  \\"layout1/js\\": {
    \\"tag\\": \\"layout1\\",
    \\"type\\": \\"js\\",
    \\"data\\": [
      \\"podlet1\\"
    ]
  }
}",
  "meta/tags.json": "{
  \\"podlet1/js\\": \\"4997b071527eee6c12db24957abc8b30b63dc46c76a2a1a04826852bf220dcd6\\"
}",
  "podlet1-4997b071527eee6c12db24957abc8b30b63dc46c76a2a1a04826852bf220dcd6.json": "[
  {
    \\"id\\": \\"c645cf572a8f5acf8716e4846b408d3b1ca45c58\\",
    \\"source\\": \\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",
    \\"deps\\": {},
    \\"file\\": \\"./assets/js/bar.js\\"
  }
]",
}
`;

exports[`integration - publish instructions before publishing assets 1`] = `
Object {
  "8b9a3142fa84751f4a0b6578f58024bd3b98aea6fb47f8e87dd0c813a250fa95.js": "\\"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\\\\\"function\\\\\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\\\\\"Cannot find module '\\\\\\"+o+\\\\\\"'\\\\\\");throw f.code=\\\\\\"MODULE_NOT_FOUND\\\\\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\\\\\"function\\\\\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({},{},[]);\\\\n\\"",
  "meta/instructions.json": "{
  \\"layout1/js\\": {
    \\"tag\\": \\"layout1\\",
    \\"type\\": \\"js\\",
    \\"data\\": [
      \\"podlet1\\"
    ]
  }
}",
  "meta/tags.json": "{
  \\"podlet1/js\\": \\"4997b071527eee6c12db24957abc8b30b63dc46c76a2a1a04826852bf220dcd6\\"
}",
  "podlet1-4997b071527eee6c12db24957abc8b30b63dc46c76a2a1a04826852bf220dcd6.json": "[
  {
    \\"id\\": \\"c645cf572a8f5acf8716e4846b408d3b1ca45c58\\",
    \\"source\\": \\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",
    \\"deps\\": {},
    \\"file\\": \\"./assets/js/bar.js\\"
  }
]",
}
`;

exports[`integration - publishing a feed via the /publish-assets endpoint 1`] = `
Object {
  "file": "podlet1-4997b071527eee6c12db24957abc8b30b63dc46c76a2a1a04826852bf220dcd6.json",
  "id": "4997b071527eee6c12db24957abc8b30b63dc46c76a2a1a04826852bf220dcd6",
}
`;
